onnxruntime/core/framework/session_state.cc

```diff
@@ -17,44 +17,91 @@ using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
 
-const GraphViewer* SessionState::GetGraphViewer() const { return graph_viewer_.get(); }
-Status SessionState::SetGraph(const Graph& graph) {
-  graph_viewer_ = onnxruntime::make_unique<onnxruntime::GraphViewer>(graph);
-  auto& logger = Logger();
+void SessionState::SetupAllocators() {
+  for (const auto& provider : execution_providers_) {
+    for (const auto& allocator : provider->GetAllocators()) {
+      const OrtMemoryInfo& memory_info = allocator->Info();
+      if (allocators_.find(memory_info) != allocators_.end()) {
+        // EPs are ordered by priority so ignore the duplicate but log a message at ERROR level as this shouldn't
+        // happen.
+        LOGS(logger_, ERROR) << "Allocator already registered for " << allocator->Info()
+                             << ". Ignoring allocator from " << provider->Type();
+      } else {
+        // slightly weird indirection to go back to the provider to get the allocator each time it's needed
+        // in order to support scenarios such as the CUDA EP's per-thread allocator.
+        allocators_[memory_info] = [&provider](int id, OrtMemType mem_type) {
```

Comment:
The previous code also had the same indirection, just in a different way (no delegates). Wondering what does this change solve then? Is this temporary until we figure out the per-thread CUDA allocator?

Reply:
Initially I didn't have a delegate and it was much nicer and of clearer value. Ideally we can get back to that by resolving the CUDA EP per-thread allocator question (or the CUDA EP can make the per-thread aspect an implementation detail of a wrapper IAllocator implementation).

This setup has one less level of indirection vs. ExecutionProviders, and given the lookup is based on OrtMemoryInfo it (to me anyway) makes more sense to have this information in SessionState as that's more consistent which where we get all the other info from. Also if I wanted to override the allocator for an OrtMemoryInfo value with a custom allocator that does not come from an EP having this lookup in ExecutionProviders doesn't make sense as I don't think we need to tightly couple a device allocator with a specific EP.

---
In reply to: [443765924](https://github.com/microsoft/onnxruntime/pull/4194#discussion_r443765924) [](ancestors = 443765924)